File: rollTable.py
#!/usr/bin/python
from random import random
import sys
p = 1
n = 1
current = None
if (len(sys.argv) == 2):
    for line in open(sys.argv[1]):
        if(len(line) > 0):
            if random() < p:
                current = line
            n += 1
            p = 1.0/n
    print(current[:-1])

File: rollTable.js
const { execSync } = require('child_process')
const fs = require('fs')

/**
 * Gets a random line from a file (eg. quotes, bullshit table, madness table)
 *
 * @param {String} fileName - name of file to get a random line from
 */
function rollTable(filename) {
	return new Promise((res, rej) => fs.readFile(filename, 'utf8', (err, data) => { if(err) rej(err); else res(data) }))
	.then(data => {
		const lines = data.split("\n").filter(s => s.length > 0)
		let n = 1, current = null
		for(let line of lines) {
			if(Math.random() < 1/n)
				current = line
			n += 1
		}
		return current
	})
}

module.exports = rollTable
